public-docs/dev-debug-app-slowness.md

```diff
@@ -0,0 +1,88 @@
+    - The `millis` field tells you how slow it was.
+    - The `planSummary` field tells you whether an index was used and which index. If it contains the string `IXSCAN`, an index was used for the lookup. `IDHACK` is also good because it means the internal `_id` index was used. If it's `COLLSCAN`, this is probably why it's slow. Think about whether the query could be rewritten to include a field that is already indexed. If not, think about whether a new index might help. Refer to [the official MongoDB indexing docs](https://docs.mongodb.com/manual/tutorial/create-indexes-to-support-queries/)
+
+5. If you've identified a slow query using `COLLSCAN`, at this point it's easiest to begin running each slow query directly in the MongoDB shell while trying various optimizations. There are two main ways to try to get MongoDB to use an index:
+    - Modify the query to add a field that is already indexed. Keep in mind that this field value must be available to your code that is running the query.
+    - Add a new index for the fields in your query. Do this [directly from the MongoDB shell](https://docs.mongodb.com/manual/reference/method/db.collection.createIndex/) for now.
+
+6. After each fix you try, rerun the offending query and then rerun the `system.profile` query to see whether that query is still slow enough to appear in the results. (If it does still appear but it's now using `IXSCAN` instead of `COLLSCAN`, that's progress. Continue with these steps to see if other improvements can be made to make it faster.) You can add additional filters to the `find` to find only a particular command or collection. It's also helpful to filter out any queries that were in your previous results. To do this, copy the `ts` field value from the first document in your results, and add it to your next query filter:
+
+    ```
+    db.system.profile.find({
+      ts: {
+        $gt: ISODate("2019-04-26T02:29:39.315+0000")
+      }
+    }).sort({ ts: -1.0 });
+    ```
+
+## Step 5: Look at other potential causes
+
+1. If a query is slow but it used an index for lookup, often the slowness is caused by the `sort`. Does the slow query have a specified sort, and if so, is there an index that supports that sort? If not, create one locally and check whether the query is faster.
+    - Tip: A sort with multiple keys can only use an index with those same keys in the same order. The index direction need not be the same as the sort direction. For example, if the query sort is `{ createdAt: 1, _id: 1 }`, Mongo will use a `{ createdAt: 1, _id: 1 }` index or a `{ createdAt: -1, _id: -1 }` index, but it will not use a `{ createdAt: 1 }` index.
+2. If a query is optimized and MongoDB is using an index for both lookup and sorting, look at what fields are being used by the code. It's always a good idea to include the `projection` option with a list of only the fields you actually need. This may not make a huge difference in the speed, but it benefits memory and bandwidth as well.
```

Comment:
Glad you mentioned this. I think we’ve got a some queries where we’re returning a lot more fields that we need.